### PR TITLE
(core) customizable default port, instance links

### PR DESCRIPTION
**Pull request by anotherchrisberry**
Adds two new fields to the application attributes: `instancePort` and `instanceLinks`. `instancePort` is used to generate links into instances in the application; `instanceLinks` provides a custom "Links" section in the instance details.

I'm not sure how this works with GCP and Kubernetes - @duftler @lwander does it make sense to include the instance port in the instance DNS links for those providers?

### CODE REVIEW

File: app/scripts/modules/core/application/modal/editApplication.controller.modal.js [hunk 1]
```diff
@@ -19,6 +20,7 @@ module.exports = angular
     vm.errorMsgs = [];
     vm.application = application;
     vm.applicationAttributes = _.cloneDeep(application.attributes);
+    vm.applicationAttributes.instancePort = vm.applicationAttributes.instancePort || settings.defaultInstancePort || 8080;
```
Comment (duftler) [inline app/scripts/modules/core/application/modal/editApplication.controller.modal.js]: Why is this `8080`?

**Comment by duftler** [comment]
I think it would make sense to include the instance port for GCE, as long as it's optional.

I didn't run the branch, but from reading it it seems like when I create a new application it'll always pick up the global default from settings.js or fall back to 7001 if that isn't defined.

So I can see how when creating a new application, I'll have to clear the instance port field if I don't want an instance port defined for that particular application. But it also seems like when I Edit an application, it'll define an instance port if I haven't defined one. Is that for migrating existing applications onto the new scheme?

I really like this whole approach, but would hope there is some easy way to explicitly not specify an instance port for a given application. Having the Edit always assign one seems to make it harder to do this. I may have misread the edit path though.

**Comment by lwander** [comment]
To clarify, does `instancePort` connect to an instance, and `instanceLink` connect to an external page showing details about the instance? For example, `instancePort` would open the service the instance is running, and `instanceLink` would open an AWS page?

File: app/scripts/modules/core/application/modal/editApplication.controller.modal.js (hunk 1 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/application/modal/editApplication.controller.modal.js]: Whoops, good catch. It should probably default to port 80 if nothing's configured.

**Comment by anotherchrisberry** [comment]
@lwander `instancePort` is the default port open on an instance, i.e. how a user would access the web server, i.e. what you said. `instanceLink`s are just links into endpoints served up by that instance, e.g. health check, metrics, logs - but they're just deep links into the instance itself.

**Comment by gbougeard** [comment]
Would it be pertinent to be able to configure 3rd parties links like a Grafana dashboard link and/or Sumologic one to have a quick link to the instance logs and/or a custom metrics dashboard?
Do you think it would make sense to have this kind of thing in Spinnaker UI?

**Comment by anotherchrisberry** [comment]
A few updates from the original cut of this:
  1. rather than configuring this when creating a new application, I've moved it to the Config tab
  2. instance port is optional and does not get a default value unless it's configured to do so in settings.js (the details view will use port 80 if no default is configured and nothing is configured in the application attributes)
  3. expanded the functionality to entire sections:
screen shot 2016-06-22 at 10 30 14 am
which results in multiple sections of links appearing in the details, e.g.
screen shot 2016-06-22 at 10 47 03 am
  4. instance links can be defaulted at the system level, i.e. in settings.js - but they won't be persisted with the application unless customization occurs

**Comment by anotherchrisberry** [comment]
@gbougeard I'm not familiar with Sumologic or Grafana, but we could probably enhance this and add some templating capabilities into the links to construct what's needed, assuming what's needed is the machine's IP address.

**Comment by gbougeard** [comment]
@anotherchrisberry :+1: the Config tab in which we can add custom links would be perfect if there was this templating capabilities. This way we could add "external" links;
this is an example of sumologic url using the private DNS : https://service.sumologic.com/ui/#section/search/@_sourcehost=%22ip-10-0-5-174.eu-west-1.compute.internal%22 
And for grafana urls it's quite the same : if it's possible to use `$instanceIp`  or '$inctancePrivateDns` or however you name it; it would be awesome

**Comment by anotherchrisberry** [comment]
@gbougeard updated to handle some interpolation. Once this is in, let's work together to figure out what those fields should be and then I can update the text to provide some context around how the interpolation works.

**Comment by gbougeard** [comment]
Do you mean `instance` attributes accessible for example as `$privateDnsName` `$publicDnsName` `$privateIpAddress` `$publicIpAddress` ?

**Comment by anotherchrisberry** [comment]
@gbougeard yep - since those fields will vary from cloud provider to cloud provider, we'll want to make whatever help contents we provide pluggable.

**Comment by anotherchrisberry** [comment]
I am going to close this and split it into separate PRs for ports and links. This has gotten much bigger than I meant for it to get.

**Comment by duftler** [comment]
I like the proposed design.

**Pull request closed without merging**